cmd/kubeturbo/app/kubeturbo_builder.go

```diff
@@ -344,3 +367,64 @@ func checkServerVersion(restClient restclient.Interface) bool {
 
 	return false
 }
+
+func discoverk8sAPIResourceGV(client *kubernetes.Clientset, resourceName string) (schema.GroupVersion, error) {
+	// We optimistically use a globally set default if we cannot discover the GV.
+	defaultGV := util.K8sAPIDeploymentReplicasetDefaultGV
+
+	apiResourceLists, err := client.ServerPreferredResources()
+	if apiResourceLists == nil {
+		return defaultGV, err
+	}
+	if err != nil {
+		// We don't exit here as ServerPreferredResources can return the resource list even with errors.
+		glog.Warningf("Error listing api resources: %v", err)
+	}
+
+	latestExtensionsVersion := schema.GroupVersion{Group: util.K8sExtensionsGroupName, Version: ""}
+	latestAppsVersion := schema.GroupVersion{Group: util.K8sAppsGroupName, Version: ""}
+	for _, apiResourceList := range apiResourceLists {
+		if len(apiResourceList.APIResources) == 0 {
+			continue
+		}
+
+		found := false
+		for _, apiResource := range apiResourceList.APIResources {
+			if apiResource.Name == resourceName {
+				found = true
+				break
+			}
+		}
+		if found == false {
+			continue
+		}
+
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			return defaultGV, fmt.Errorf("Error parsing GroupVersion: %v", err)
+		}
+
+		group := gv.Group
+		version := gv.Version
+		if group == util.K8sExtensionsGroupName {
+			latestExtensionsVersion.Version = latestComparedVersion(version, latestExtensionsVersion.Version)
+		} else if group == util.K8sAppsGroupName {
+			latestAppsVersion.Version = latestComparedVersion(version, latestAppsVersion.Version)
+		}
+	}
+
+	if latestAppsVersion.Version != "" {
+		return latestAppsVersion, nil
+	}
+	if latestExtensionsVersion.Version != "" {
+		return latestExtensionsVersion, nil
+	}
```

Comment:
Does this logic mean that we always prefer appsVersion over extensionsVersion, or it is not possible for a resource to belong to both appsGroup and extensionsGroup?

Reply:
So the evolution of API resources from `extensions` to respective other groups has not been very elegant. There were many resources which were all collectively listed under group `extensions` and moved into their respective groups in the later releases, `deployments` moved to `apps` and `ingress` moved to `networking.k8s.io` being few examples.

It is possible for a resource to exist under multiple groups.
The strategy used here (which is kind of hard coding) is to simply prefer `apps` over `extensions`.

In case we move our code such that all other resource versions need to be discovered (I don't see the need for that as yet); a simple strategy would be to always not prefer choosing `extentions` if a resource is found in multiple groups.

Reply:
Thanks for the info. That makes sense to me.